File: app/src/main/java/com/ternaryop/photoshelf/parsers/AndroidTitleParserConfig.java
package com.ternaryop.photoshelf.parsers;

import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import android.content.Context;

import com.ternaryop.photoshelf.util.AssetsJsonConfig;
import org.json.JSONObject;

/**
 * Created by dave on 21/03/15.
 * The Config to use under Android, read the config from assets and upgrade it if an imported version exists
 */
public class AndroidTitleParserConfig extends AssetsJsonConfig implements TitleParserConfig {
    private static final String TITLE_PARSER_FILENAME = "titleParser.json";

    private static int version = -1;
    private static final JSONTitleParserConfig titleParserConfig = new JSONTitleParserConfig();

    public AndroidTitleParserConfig(Context context) {
        if (version > 0) {
            return;
        }

        try {
            final JSONObject jsonObject = readAssetsConfig(context, TITLE_PARSER_FILENAME);
            version = readVersion(jsonObject);
            titleParserConfig.readConfig(jsonObject);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public int getVersion() {
        return version;
    }

    @Override
    public List<TitleParserRegExp> getTitleCleanerList() {
        return titleParserConfig.getTitleCleanerList();
    }

    @Override
    public String applyList(List<TitleParserRegExp> titleParserRegExpList, String input) {
        return titleParserConfig.applyList(titleParserRegExpList, input);
    }

    @Override
    public Pattern getTitleParserPattern() {
        return titleParserConfig.getTitleParserPattern();
    }

    @Override
    public List<LocationPrefix> getLocationPrefixes() {
        return titleParserConfig.getLocationPrefixes();
    }

    @Override
    public Map<String, Pattern> getCities() {
        return titleParserConfig.getCities();
    }
}

File: app/src/main/java/com/ternaryop/photoshelf/parsers/AndroidTitleParserConfig.kt
package com.ternaryop.photoshelf.parsers

import android.content.Context
import java.util.regex.Pattern

/**
 * Created by dave on 21/03/15.
 * The Config to use under Android, read the config from assets and upgrade it if an imported version exists
 */
private const val TITLE_PARSER_FILENAME = "titleParser.json"

class AndroidTitleParserConfig(context: Context) : AssetsJsonConfig(), TitleParserConfig {

    override val titleCleanerList: List<TitleParserConfig.TitleParserRegExp>
        get() = titleParserConfig.titleCleanerList

    override val titleParserPattern: Pattern
        get() = titleParserConfig.titleParserPattern

    override val locationPrefixes: List<LocationPrefix>
        get() = titleParserConfig.locationPrefixes

    override val cities: Map<String, Pattern>
        get() = titleParserConfig.cities

    override val version: Int = _version

    init {
        if (version <= 0) {
            try {
                val jsonObject = readAssetsConfig(context, TITLE_PARSER_FILENAME)
                _version = readVersion(jsonObject)
                titleParserConfig.readConfig(jsonObject)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun applyList(titleParserRegExpList: List<TitleParserConfig.TitleParserRegExp>, input: String): String {
        return titleParserConfig.applyList(titleParserRegExpList, input)
    }

    companion object {
        private var _version = -1
        private val titleParserConfig = JSONTitleParserConfig()
    }
}
